### PR TITLE
Da#219 create cluster overview window

### PR DESCRIPTION
**Pull request by AayushTyagi1**
Get Cluster overview Information

### CODE REVIEW

**Review by AayushTyagi1 — COMMENTED** [review]
Nice Work @lokesh-couchbase

1. If the status is 'Healthy' we can show it in green else red. 
2. Node full attribute should be visible when hover over it
3. Services should be Search, Index, Data, Query rather than fts, index, kv, n1ql 
4. Up Time: 2491 should be in days, hours and minutes

**Comment by lokesh-couchbase** [comment]
> Nice Work @lokesh-couchbase
> 
> 1. If the status is 'Healthy' we can show it in green else red.
> 2. Node full attribute should be visible when hover over it
> 3. Services should be Search, Index, Data, Query rather than fts, index, kv, n1ql
> 4. Up Time: 2491 should be in days, hours and minutes

Resolved the issues

File: src/pages/overviewCluster/overviewCluster.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    };
+
+    try {
+        // Fetch server overview details
+        const restAPIObject = new CouchbaseRestAPI(connection);
+        const serverOverview: ServerOverview | undefined = await restAPIObject.getOverview();
+
+        // Fetch Buckets
+        let bucketsSettings = await connection?.cluster?.buckets().getAllBuckets();
+        let Buckets = fetchBucketNames(bucketsSettings, connection);
```
Comment (AayushTyagi1) [inline src/pages/overviewCluster/overviewCluster.ts:72]: variables should be in camelCase keep it buckets

File: src/pages/overviewCluster/overviewCluster.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        // Buckets Data
+        let bucketsHTML: IKeyValuePair[] = [];
+        for (let bucket of Buckets) {
+            const bucketOverview: BucketOverview | undefined = await restAPIObject.getBucketsOverview(bucket.name);
+            let bucketHTML = bucketOverview !== undefined ? getBucketData(bucketOverview) : '';
+            bucketsHTML.push({ key: bucket.name, value: bucketHTML });
+        }
+
+        // Nodes Data
+        const NodesHTML: IKeyValuePair[] = serverOverview !== undefined ? getNodeTabData(serverOverview) : [];
```
Comment (AayushTyagi1) [inline src/pages/overviewCluster/overviewCluster.ts:89]: Keep it nodesHTML

File: src/pages/overviewCluster/overviewCluster.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        ClusterOverviewObject = {
```
Comment (AayushTyagi1) [inline src/pages/overviewCluster/overviewCluster.ts:91]: camelCase problem please fix
Keep it clusterOverviewObject
and fix whole object

File: src/types/IClusterOverview.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ *   limitations under the License.
+ */
+import { Bucket } from "couchbase";
+import { BucketDetails, OverviewGeneral } from "../util/OverviewClusterUtils/OverviewClusterHelper";
+import { IKeyValuePair } from "./IKeyValuePair";
+import { CBNode } from "../util/apis/CBNode";
+
+
+export interface IClusterOverview {
+  readonly GeneralDetails: OverviewGeneral | null;
```
Comment (AayushTyagi1) [inline src/types/IClusterOverview.ts:23]: generalDetails
Follow the same in others

File: src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                    ${kv.value} \
+                </div> \
+            </div> \
+        `)).join('')} \
+    </div> \
+    `;
+};
+
+export const getNodeTabData = (serverOverview: ServerOverview): IKeyValuePair[] => {
+    let NodesData: IKeyValuePair[] = [];
```
Comment (AayushTyagi1) [inline src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts:327]: change to nodesData

File: src/util/OverviewClusterUtils/OverviewClusterHelper.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
+import { IKeyValuePair } from "../../types/IKeyValuePair";
+
+export class OverviewGeneral {
+    public Cluster: IKeyValuePair[] | null;
```
Comment (AayushTyagi1) [inline src/util/OverviewClusterUtils/OverviewClusterHelper.ts:4]: Keep the variable in camelCase like cluster quota

File: src/util/OverviewClusterUtils/OverviewClusterHelper.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        this.Storage = Storage;
+      }
+}
+
+export class BucketDetails {
+  public Bucket: IKeyValuePair[] | null;
+  public Quota: IKeyValuePair[] | null;
+  public BasicStats: IKeyValuePair[] | null;
+  constructor(
+    Bucket: IKeyValuePair[] | null,
```
Comment (AayushTyagi1) [inline src/util/OverviewClusterUtils/OverviewClusterHelper.ts:26]: camelCase for bucket, quota and basicStats

File: src/pages/overviewCluster/overviewCluster.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import { ServerOverview } from "../../util/apis/ServerOverview";
+import { IKeyValuePair } from "../../types/IKeyValuePair";
+import { BucketOverview } from "../../util/apis/BucketOverview";
+import { fmtByte, formatServices, mbToGb } from "../../util/OverviewClusterUtils/OverviewClusterHelper";
+import { getNodeTabData } from "../../util/OverviewClusterUtils/ClusterOverviewNodeTab";
+import { getBucketData } from "../../util/OverviewClusterUtils/ClusterOverviewBucketTab";
+import { getGeneraStorageDetails, getGeneralClusterDetails, getGeneralQuotaDetails, getGeneralRAMDetails } from "../../util/OverviewClusterUtils/ClusterOverviewGeneralTab";
+
+const fetchBucketNames = (bucketsSettings: BucketSettings[] | undefined, connection: IConnection): Array<Bucket> => {
+    let Buckets: Array<Bucket> = [];
```
Comment (AayushTyagi1) [inline src/pages/overviewCluster/overviewCluster.ts:21]: camelCase

File: src/pages/overviewCluster/overviewCluster.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            enableScripts: true,
+            enableForms: true,
+        }
+    );
+
+    currentPanel.iconPath = vscode.Uri.file(
+        path.join(__filename, "..", "..", "images", "cb-logo-icon.svg")
+    );
+    currentPanel.webview.html = `Loading....`;
+    let ClusterOverviewObject: IClusterOverview = {
```
Comment (AayushTyagi1) [inline src/pages/overviewCluster/overviewCluster.ts:56]: Keep it clusterOverviewObject

File: src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts [hunk 10]
```diff
@@ -0,0 +1,338 @@
+import { IKeyValuePair } from "../../types/IKeyValuePair";
+import { fmtByte, formatDuration, formatNumber, formatServices } from "./OverviewClusterHelper";
+import { CBNode } from "../apis/CBNode";
+import { ServerOverview } from "../apis/ServerOverview";
+import { Constants } from "../constants";
+
+const getNodeTabDetails = (NodeDetails: CBNode | undefined): IKeyValuePair[] => {
+    let details: IKeyValuePair[] = [];
+    if (NodeDetails === undefined) {
```
Comment (AayushTyagi1) [inline src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts:9]: Keep it nodeDetails

Comment (AayushTyagi1) [reply src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts:9]: Do it for other functions too

Comment (lokesh-couchbase) [reply src/util/OverviewClusterUtils/ClusterOverviewNodeTab.ts:9]: Updated for all functions in the file

File: src/pages/overviewCluster/overviewCluster.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                Quota: generalQuotaDetails,
+                Storage: generalStorageDetails,
+                RAM: generalRAMDetails,
+            },
+            BucketsHTML: bucketsHTML,
+            NodesHTML: NodesHTML,
+        };
+    } catch (err) {
+        logger.error("Failed to get Cluster Overview Data, error: " + err);
+        currentPanel.webview.html = `Error while loading cluster, Please try again later!`;
```
[the remark this answers is not in the document]
Comment (lokesh-couchbase) [reply src/pages/overviewCluster/overviewCluster.ts]: Using this Modal to show error message now

**Comment by AayushTyagi1** [comment]
Approved

**Pull request merged**